### PR TITLE
#Module4

### PR DESCRIPTION
**Pull request by TheElidanXD**
Add Groups.
Add many-to-many relation between Users and Groups.

### CODE REVIEW

File: module3/src/controllers/group.controller.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+export const groupRouter = express.Router();
+groupRouter.use(express.json());
+
+groupRouter.get('/:id', (req, res) => {
+    groupService.getGroupById(req.params.id)
+        .then((result: unknown) => {
+            res.json(result);
+        })
+        .catch((err: Error) => {
+            res.status(404).json(err);
```
Comment (evgenDer) [inline module3/src/controllers/group.controller.ts]: you can use HttpStatuse from @types/node

File: module3/src/controllers/group.controller.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        });
+});
+
+export const groupRouter = express.Router();
+groupRouter.use(express.json());
+
+groupRouter.get('/:id', (req, res) => {
+    groupService.getGroupById(req.params.id)
+        .then((result: unknown) => {
```
Comment (evgenDer) [inline module3/src/controllers/group.controller.ts]: why is result unknown?

File: module3/src/validators/group.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import Joi from 'joi';
+import { availableGroupPermission } from '../models/group';
+
+export const groupSchema = Joi.object().keys({
+    name: Joi.string().required(),
+    permissions: Joi.array()
+        .items(
+            Joi.string().custom((value: string, helpers) => {
+                return availableGroupPermission.includes(value) ?
```
Comment (evgenDer) [inline module3/src/validators/group.ts]: you can use || - `availableGroupPermission.includes(value) || helpers.error(`There is no such permission as ${value}`)`

File: module3/src/controllers/group.controller.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        });
+});
+
+groupsRouter.post('/', validateSchema(groupSchema), (req, res) => {
+    groupService.createGroup(req.body)
+        .then((result: unknown) => {
+            res.json(result);
+        })
+        .catch((err: Error) => {
+            res.status(404).json(err);
```
Comment (evgenDer) [inline module3/src/controllers/group.controller.ts]: duplicate code, read about next callback

File: module3/src/services/user.service.ts [hunk 5]
```diff
@@ -1,23 +1,21 @@
...
+    private user = UserModel(db);
+    private userGroup = UserGroupModel(db);
 
     getAllUsers = () => this.user.findAll(
         { where: { isdeleted: false } }
     );
 
-    createUser = (newUser: Omit<unknown, string>) => this.user.create(newUser);
+    createUser = (newUser: User) => {
+        newUser.id = uuid();
```
Comment (evgenDer) [inline module3/src/services/user.service.ts]: you can add this option from db side

**Review by evgenDer — CHANGES_REQUESTED** [review]

File: module3/src/validators/group.ts (hunk 3 above)
Comment (TheElidanXD) [reply module3/src/validators/group.ts]: Done

File: module3/src/controllers/group.controller.ts (hunk 2 above)
Comment (TheElidanXD) [reply module3/src/controllers/group.controller.ts]: Created types for group, user and userGroup instances

File: module3/src/services/user.service.ts (hunk 5 above)
Comment (TheElidanXD) [reply module3/src/services/user.service.ts]: Done

File: module3/src/controllers/group.controller.ts (hunk 4 above)
Comment (TheElidanXD) [reply module3/src/controllers/group.controller.ts]: Done

File: module3/src/controllers/group.controller.ts (hunk 1 above)
Comment (TheElidanXD) [reply module3/src/controllers/group.controller.ts]: Added StatusCodes enum

**Pull request merged**